pm, server: close RTMP connection when sender validation fails

kyriediculous opened the pull request:
**What does this pull request do? Explain your changes. (required)**
<!-- A clear and concise description of what this pull request does. -->

**Specific updates (required)**
- Return a top-level error from `sender.validateSender` -> `pm.ErrSenderValidation` so that we can compare a returned error against it in the `server` package
- Introduce a `server.nonRetryableTranscodeError` type that implements the `error` interface
- Wrap `sessions == nil` and failing sender validation in `nonRetryableTranscodeError``
- Modify `server.shouldStopStream(err)` to return whether `err` is a `nonRetryableTranscodeError`
- Close RTMP stream when `server.shouldStopStream(err)` return true and break out of the loop in `server.processSegment`
 
**How did you test each of these updates (required)**
<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->


**Does this pull request close any open issues?**
Fixes #1247

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] README and other documentation updated
- [ ] Node runs in OSX and devenv
- [ ] All tests in `./test.sh` pass

Commit:
pm: ErrSenderValidation type

Commit:
server: close RTMP connection when a sender validation error occurs

yondonfu commented on pm/sender.go:
```diff
@@ -97,7 +100,7 @@ func (s *sender) CreateTicketBatch(sessionID string, size int) (*TicketBatch, er
 
 	if err := s.validateSender(); err != nil {
-		return nil, err
+		return nil, ErrSenderValidation
```
We should make sure that the underlying error returned by `s.validateSender()` is not dropped here.

Instead of returning a specific error here, what if we returned a custom error type? Something like:

```
type SenderValidationError struct {
    error
}

...

func (s *sender) CreateTicketBatch(...) (...) {
    ...

    if err := s.validateSender(); err != nil {
        return nil, &SenderValidationError{err}
    }
}
```

If we use a custom error type, then the underlying error message can still be logged by the top level caller. The caller can also do a type assertion to check the error type.

yondonfu commented on server/broadcast.go:
```diff
@@ -276,9 +276,20 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) error {
 		}
+
+		// non retryable error, close stream
+		if shouldStopStream(err) {
```
The updated impl of `shouldStopStream()` would close the RTMP conn when there are no sessions available for a segment which might not be desirable. I think there is a difference between a non-retryable error that closes the RTMP conn vs. a non-retryable error that allows the stream to continue (but drops the current segment). We currently skip the segment when there are no sessions available, but do not close the RTMP conn.

yondonfu commented on server/broadcast.go:
```diff
@@ -330,11 +337,6 @@ func transcodeSegment(cxn *rtmpConnection, seg *stream.HLSSegment, name string)
 			if res == nil && err == nil {
 				return errors.New("Empty response")
 			}
-			if shouldStopStream(err) {
```
Any particular reason why this code block had to move up to `processSegment()`?

kyriediculous replied:
If I'm not mistaken this code block didn't really surve a purpose before as `shouldStopStream()` would just always return false and so the code block would never be executed. Therefore I just decided to reuse it for the purpose of checking non-retryable transcode errors.

yondonfu commented on server/errors.go:
```diff
@@ -0,0 +1,13 @@
+package server
+
+type nonRetryableTranscodeError struct {
```
Most of this file could be omitted if we embed `error` in the struct i.e.

```
type nonRetryableTranscodeError struct {
    error
}
```

j0sh replied:
Is this file still needed? I don't see `nonRetryableTranscodeError` used anywhere else in this PR

kyriediculous commented:
Ready for full review

kyriediculous commented:
Nope that's right. I thought i had removed it in a previous commit, removed
now.

On Fri, Jan 24, 2020 at 6:31 PM Josh Allmann <notifications@github.com>
wrote:

> *@j0sh* commented on this pull request.
> ------------------------------
>
> In server/errors.go
> <https://github.com/livepeer/go-livepeer/pull/1318#discussion_r370753559>:
>
> > @@ -0,0 +1,13 @@
> +package server
> +
> +type nonRetryableTranscodeError struct {
>
> Is this file still needed? I don't see nonRetryableTranscodeError used
> anywhere else in this PR
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/livepeer/go-livepeer/pull/1318?email_source=notifications&email_token=AFJZZWVLO7WYMZEF3DYEK6DQ7MQYJA5CNFSM4KKLD5EKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCS735WI#discussion_r370753559>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFJZZWVG65CMIHSRM4GYHJ3Q7MQYJANCNFSM4KKLD5EA>
> .
>

yondonfu commented on server/broadcast.go:
```diff
@@ -276,16 +277,25 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) error {
 			return nil
 		}
+
+		// non retryable error, close stream
```
At the moment, we only close the stream for certain non-retryable errors. We don't close the stream for other non-retryable errors (i.e. "No sessions available"). I'd be ok with omitting the comment here entirely since the name of `shouldStopStream()` seems self-explanatory.

yondonfu commented on pm/sender.go:
```diff
@@ -9,6 +9,11 @@ import (
 	"github.com/pkg/errors"
 )
 
+// ErrSenderValidation is emitted when sender.validateSender() encounters an error
```
In case `ErrSenderValidation` is ever returned from a function that is not called `validateSender()` (i.e. if there was a refactor), we could drop `sender.validateSender()` from this comment and just say something like:

```
ErrSenderValidation is returned when the sender cannot send tickets
```

yondonfu reviewed (changes requested):
Ah looks like the function signature of `transcodeSegment()` has changed on `master` (changes from [multi-part return PR](https://github.com/livepeer/go-livepeer/commit/68fdf092377740d34b73f7108fe15e5dc8bbb3d4#diff-cecc27bfe2b6b19181ce443733dfba57)) and `shouldStopStream()` was removed. Think its worth rebasing both `nv/b-deposit-withdrawround-check` on top of `master` to make sure that the updates in this branch are compatible.

kyriediculous commented:
fixed requested changes and rebased on top of master

yondonfu reviewed (approved):
LGTM
